CODING_GUIDELINES.md

```diff
@@ -1,37 +1,186 @@
-- No hard-coded colors, always use the existing helpers and variables.
-- Always respect our baseline for vertical rhythm.
+## Naming conventions
 
-## HTML templates
+### General guidelines
 
-- Any text displayed to the user should be projected, never received from inputs or attributes. 
-This always allows better customizability, and prevents blocking issues (or even security issues)
-if said text can contain HTML tags or interpolation.
+- Prefer full words over abbreviations
+- Anything public and meant to be used by consumers should have the `clr` prefix.
 
-## Typescript code
+### SCSS
+
+- SCSS variables should use the `clr` prefix and follow the `kebab-case` convention: `$clr-some-variable`.
+- Coming soon...
+
+### Typescript
+#### Constants
+Constants internal to Clarity are named following the `SCREAMING_SNAKE_CASE` convention.
+Constants exported as part of our public API follow the same convention but need to be prefixed: `CLR_SCREAMING_SNAKE_CASE`.
+
+#### Enums
+The enum itself is prefixed by `Clr` since it is public, and each of the values are `SCREAMING_SNAKE_CASE` because they are constant. For instance:
+```typescript
+ClrDirection.DOWN
+```
+
+#### Class names
+Any class exported publicly needs to be prefixed with `Clr`. For instance:
+```typescript
+class ClrBourgeoisie {}
+class ClrProletariat {}
+```
+
+#### Properties and methods
+Properties and methods on Typescript classes should not be prefixed or suffixed. In particular, private properties should **not** be prefixed with `_`. 
+```typescript
+public open: boolean;
+public next() { ... }
+private subscription: Subscription;
+```
+
+There are 2 exceptions to this rule, and they're the only cases where we allow the use of an `_` prefix:
+
+- A private property is hidden behind a getter (and maybe a setter) that has the same name:
+```typescript
+private _open: boolean;
+public get open(): boolean;
+public set open(value: boolean);
+
+private _openChange: Subject<boolean>;
+public get openChange(): Observable<boolean>;
+```
+- A property or method needs to be public for integration purposes but should not be considered public by the consumer:
+```typescript
+@ContentChildren(Something) _children: QueryList<Something>;
+_internalMethodCalledFromAnotherClass()
+```
+
+
+### Angular
+#### Components and directives selector
+All Angular selectors are prefixed with `clr`, then follow the typical `snake_case` or `camelCase` depending on whether they're element or attribute selectors:
+```
+clr-component
+[clrDirective]
+```
+When larger components have sub-components, the sub-component should use these selectors:
+```
+clr-component-subcomponent
+[clrComponentSubdirective]
+```
+This is slightly verbose, but will prevent conflicts between components. This is a case where abbreviations are acceptable to reduce verbosity.
+
+#### Inputs and outputs
+To avoid conflict with non-Clarity components, inputs and outputs of our directives and components should be prefixed with `clr`:
```

Comment:
Should add a statement about using `Change` vs `Changed`

Reply:
That's not a coding guideline on our side, that's a requisite from Angular if you want two-way binding. But I can add it, sure.